Add `check_pub_error` macro

tcharding opened the pull request:
Builds on top of #1742

Public errors are part of the public API, as such we need to be careful not to change them. One thing that is hard to notice is adding/removing `Send`/`Sync`.
    
Add a macro that asserts that an error type implements `Send` and `Sync`. While we are at it, assert `Debug` is implemented as well because we would like all errors to implement `Debug`.

Based on idea by @Kixunil: https://github.com/rust-bitcoin/rust-bitcoin/pull/1127#discussion_r932115136

tcharding commented:
I'm not sure why the release job `cargo publish -p bitcoin --dry-run ` fails to find the new macro?

tcharding commented:
All the force pushes are me rebasing on #1742 trying to get it past CI.

sanket1729 commented on internals/src/error.rs:
```diff
@@ -27,3 +27,15 @@ macro_rules! write_err {
+
+/// Trait used internally to assert that error types implement `Send`, `Sync`, and `Debug`.
+#[doc(hidden)]
+pub trait AssertSendSyncDebug: Send + Sync + core::fmt::Debug {}
```
Why is this pub and why do we need a trait? Any drawbacks of creating this in `#cfg(test)` for this? We can also do this with function without the trait. 

```rust
#[cfg(test)
fn assert_send_sync<T: Send + Sync + Debug> {}

macro_rules check_pub_error {
	#[cfg(test)]
	assert_send_sync<$type>();	
}
```

Kixunil replied:
Could be `pub(crate)` I think.

tcharding replied:
I ended up using a trait still but feature guarding so the trait is not built in in non-test builds. Ok?

sanket1729 replied:
Sure. I am okay as long it is only in tests.

Kixunil commented:
> assert `Debug` is implemented

We should just check for `std::error::Error` on `std`.

apoelstra commented:
Hmm, shitty.

So @tcharding the reason CI is failing is that `cargo publish` tries to use the version of `internals` that's already on crates.io. So if we add literally anything to `internals` ... or to `hashes` ... then the release job will fail.

On the one hand, this is exactly what we want to happen (it's telling us that we can't release bitcoin without releasing internals fist). But on the other hand, we're about 48 hours into this release and we're already getting a failure from the CI job, so I guess we've gotta figure out how to get it to be yellow instead of red :)

apoelstra commented:
@tcharding could you try adding a commit which does that, to see if it makes CI pass?

I think that wrapping the publish checks in `if git diff --quiet */Cargo.toml; then` ... `fi` will do it.

tcharding commented:
Spent more time playing with bash than I wanted to, will try again tomorrow.

Kixunil commented on internals/src/error.rs:
```diff
@@ -26,3 +26,27 @@ macro_rules! write_err {
+/// Trait used internally to assert that error types implement `Send`, `Sync`, and `Error`.
+#[cfg(all(test, feature = "std"))]
+#[doc(hidden)]
+pub trait AssertTraits: Send + Sync + std::error::Error {}
```
Perhaps we should also check that `no_std` errors are still `Send + Sync`?

Kixunil reviewed (commented):
Concept ACK.

apoelstra reviewed (dismissed):
ACK 928635ded299622bb08436fc44fb7028b0377b3a

stevenroose commented on internals/src/error.rs:
```diff
@@ -26,3 +26,27 @@ macro_rules! write_err {
+/// Trait used internally to assert that error types implement `Send`, `Sync`, and `Error`.
+#[cfg(all(test, feature = "std"))]
+#[doc(hidden)]
+pub trait AssertTraits: Send + Sync + std::error::Error {}
```
Still not super happy this is pub.

I would suggest the following changes:
- have a single macro and have a `($type:ty) => {};` rule for non-test, non-empty rule for test
- do Sanket's approach for methods, have one method `fn check_sendsync<T: Send + Sync>()` and `fn check_error<T: std::error::Error>()` and gate the second on std.

Currently there is a bug that non-std test runs will not be able to build.

Kixunil replied:
I'm not happy about it either but seems less work for compiler (and compile times!). It's hidden which is a clear signal people shouldn't use it (and internals is also such signal). If anyone is insane enough to depend on it they deserve the consequences.

tcharding replied:
I used you suggestions @stevenroose except I was unable to put an attribute on the arms of the macro so I kept the dummy macro for not test builds.

stevenroose reviewed (changes requested):
Left a suggestion.

tcharding commented:
Now uses suggestion by @sanket1729 to use a function instead of a trait. I also checked that the macro is used for all errors (I grepped for `std::error::Error` so ironically if any error types are missing this impl then I'll have missed them).

apoelstra commented:
I'm a little surprised that this works now without the functions being `pub`

apoelstra reviewed (dismissed):
ACK 70054bcffa51a43e60fb11f0fb99035418b1c7f1

tcharding commented:
> I'm a little surprised that this works now without the functions being `pub`

I was surprise also, I tried to work out why it works and I cannot. (Also the comment on `dead_code` is wrong, will fix it and force push.)

tcharding commented:
We should understand _why_ this actually works with privately scoped helper functions before merging this.

tcharding commented:
Face palm - thanks @Kixunil, consider me schooled.

apoelstra commented:
Oof, ditto.

tcharding commented:
I could not get it to work with functions, reverted to using traits.

apoelstra commented:
Now I'm worried again that the `impl`s are not actually taking effect when this macro is called from the downstream `bitcoin` crate. I think so because the `#[cfg(test)]` is in the macro so it should apply to the calling crate.

BTW @tcharding it occurs to me that to get privacy, the macro should define a whole `#[cfg(test)] mod internals_trait_checks { ... }` and put everything in there, the traits and their impls.

Kixunil commented on internals/src/error.rs:
```diff
@@ -26,3 +26,23 @@ macro_rules! write_err {
+        impl internals::error::AssertSendSync for $t {}
+        #[cfg(test)]
+        #[cfg(feature = "std")]
+        impl internals::error::AssertDebug for $t {}
```
Why is `Debug` gated on `std`? Seems wrong, all error types should impl `Display` and `Debug` and, in case of `std`, `std::error::Error`.

tcharding replied:
My bad, sorry for wasting review time.

Kixunil commented on internals/src/error.rs:
```diff
@@ -26,3 +26,23 @@ macro_rules! write_err {
+#[macro_export]
+macro_rules! check_pub_error {
+    ($t:ty) => {
+        #[cfg(test)]
```
It just occurred to me that I may be wrong and this doesn't work either. I would've checked but need to go rn, so I suggest you check it (by commenting-out an impl) when addressing my other comment.

tcharding commented:
I had another go, note `bitcoin::string::Error` has a generic on it that is now correctly handled. Changes in `internals::error` will need total re-review please.

Kixunil commented on internals/src/error.rs:
```diff
@@ -26,3 +26,31 @@ macro_rules! write_err {
 }
+
+/// This module is used internally to check that error types implement certain traits.
+#[cfg(feature = "trait-checks")]
```
Maybe we should use bare `cfg`, not feature, so this doesn't get pulled in when `--all-features` is used?

tcharding replied:
I used `#[cfg(trait_checks)]` and I enabled it in all builds by exporting `RUSFLAGS` from the top level `contrib/test.sh` script.

Kixunil replied:
We shoud have at least one build without it in case something gets messed up.

tcharding replied:
Oh yes, very good point.

Kixunil commented on internals/src/error.rs:
```diff
@@ -26,3 +26,31 @@ macro_rules! write_err {
+            impl<$($gen),*> AssertSendSync for $ty<$($gen),*> where $($gen: Send + Sync ),* {}
+
+            #[cfg(feature = "std")]
+            impl<$($gen),*> AssertError for $ty<$($gen),*> where $($gen: std::error::Error + 'static),* {}
```
Could also check `Debug + Display` for `no_std`.

tcharding replied:
Done as a separate trait.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -619,6 +621,8 @@ pub enum HiddenNodes {
     HiddenParts(NodeInfo),
 }
+#[cfg(feature = "trait-checks")]
+internals::check_pub_error!(HiddenNodes);
```
Unrelated but it looks like this should be called `HiddenNodesError` or be withing an `error` module.

tcharding replied:
Noted, thanks.

tcharding commented:
Changes in force push:
- Use `#[cfg(trait_checks)]` instead of feature
- Check for `Display`/`Debug` for all builds

tcharding commented:
Changes in force push:

- Changed cfg option to be `check_traits` (instead of `trait_checks`) so its a verb in line with "bench" and "bitcoin_fuzz".
- Changed CI to explicitly call `cargo build` with `RUSTFLAGS` set instead of setting for all builds.

tcharding commented:
ooo, I had missed all the public errors that were structs (as opposed to enums). Now included.

tcharding commented:
Sweeet, macro finds to unusual error types, they were added in commit: `1a2cf2681 Implement consensus encoding adapter for serde` by you @Kixunil. Are we happy with `DecodeError` and `DecodeInitError` not implementing the usual traits?

I've added `FIXME` to mark the two errors.

Kixunil commented:
Good question, IIRC those errors are meant to be converted to `serde` errors. But adding the impls wouldn't be too bad. I think we should check `Send + Sync` at least.

tcharding commented:
I elected to implement `Display` and `std::error::Error` for both types, added as an initial patch.

Kixunil commented on bitcoin/src/consensus/serde.rs:
```diff
@@ -104,10 +104,32 @@ pub mod hex {
 
+    impl fmt::Display for DecodeInitError {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "creation of hex string decoder failed")
```
Should be `write_err`.

tcharding replied:
My bad, thanks.

Kixunil commented on bitcoin/src/consensus/serde.rs:
```diff
@@ -104,10 +104,32 @@ pub mod hex {
 
+    impl fmt::Display for DecodeError {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "hex string contains invalid characters")
```
`write_err`

tcharding replied:
I'm not totally sure on the best content to put in the error message of `write_err`, noting it here but can discuss on #1862.

Kixunil commented on internals/src/error.rs:
```diff
@@ -26,3 +26,35 @@ macro_rules! write_err {
+
+/// This module is used internally to check that error types implement certain traits.
+#[cfg(check_traits)]
+#[doc(hidden)]
```
Maybe the module shouldn't be hidden but impls should be? Or at least macro should be visible, I think.

tcharding replied:
Interesting, by adding
```
[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs", "--cfg", "check_traits"]
```
and removing line 32 above, we get `check_pub_error` documented in the HTML at the crate root. And all the traits are only viewable if one goes into docs for `error` module and then `check_traits` module - seems like a good balance.

tcharding replied:
Oh I didn't mention why I left the impls unhidden, looking at the HTML it is informative to see the traits because it shows really clearly what the macro does. I don't see any benefit in hiding them.

tcharding commented:
I rebased this one, but to do the rebase properly I have to go through the whole codebase and check no new errors were introduced since I last did this. I can't be bothered doing that. Can we just agree that this PR might miss adding the macro to some errors, and keep an eye out?

sanket1729 reviewed (approved):
ACK c406cc2543acc9da3a2e0f1f270482331a2ecbca. This looks great. Did not check whether all parts are covered

apoelstra commented:
Honestly now I'm a little tempted to scrap this PR in favor of just using some creative `grep`ing on the output of #1880, which would be resilient against new errors being introduced.

tcharding commented:
That's not a bad idea because it would put further "coding policy" things into the same place instead of having a bunch of macros that one must remember to call (assuming we come up with more things like that introduced here).

tcharding commented:
Draft until #1880 resolves.

tcharding commented:
Looks like we are going to use `cargo public-api` to solve this problem.

ref: https://github.com/rust-bitcoin/hex-conservative/pull/34

Kixunil commented:
@tcharding it'd still be nice to have some tool to check that the API is good in the first place (e.g. an error type implements `std::error::Error`). `cargo public-api` helps with the biggest footguns though.

apoelstra commented:
@Kixunil right -- after bringing in `cargo public-api` we had a vague plan to write some shell scripts (or Perl or whatever) that would run in CI and do sanity checks like "if a type has `Error` in its name then it impls `std::error::Error`".

Unfortunately every specific rule we've thought of has had a bunch of exceptions :) but if you have ones in mind it'd be awesome to start doing it. Maybe in hex-conservative first where we have a much smaller API.
